torch/utils/data/standard_pipes.ipynb

```diff
@@ -0,0 +1,1006 @@
+      "\u001b[0;32m~/dataset/pytorch/torch/utils/data/datapipes/iter/grouping.py\u001b[0m in \u001b[0;36m__iter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    247\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    248\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mguaranteed_group_size\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mbiggest_size\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mguaranteed_group_size\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop_remaining\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 249\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Failed to group items'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbuffer\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mbiggest_key\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    250\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    251\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mguaranteed_group_size\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mbiggest_size\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mguaranteed_group_size\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
+      "\u001b[0;31mException\u001b[0m: ('Failed to group items', '[(0, 0)]')"
+     ]
+    }
+   ],
+   "source": [
+    "dp = ExampleIterPipe(15).map(lambda x: (x % 3, x)).shuffle().groupby(lambda x: x[0], guaranteed_group_size = 2, buffer_size = 6)\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
+  },
+  {
+   "source": [
+    "To avoid this error and drop incomplete groups, use `drop_remaining` argument"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 47,
+   "metadata": {},
+   "outputs": [
+    {
+     "output_type": "stream",
+     "name": "stdout",
+     "text": [
+      "[(2, 5), (2, 11), (2, 14), (2, 2)]\n",
+      "[(1, 7), (1, 10), (1, 13), (1, 4), (1, 1)]\n",
+      "[(0, 9), (0, 0), (0, 6), (0, 3), (0, 12)]\n"
+     ]
+    }
+   ],
+   "source": [
+    "dp = ExampleIterPipe(15).map(lambda x: (x % 3, x)).shuffle().groupby(lambda x: x[0], guaranteed_group_size = 2, buffer_size = 6, drop_remaining = True)\n",
+    "for i in dp:\n",
+    "    print(i)"
+   ]
```

Comment:
`drop_remaining` can be useful only at the end of iteration.
When the source DataPipe has not depleted yet, and we have multiple groups smaller than `guaranteed_group_size` taking up the buffer, we would encounter a problem that we don't know which group should be dropped. Should we raise RuntimeError or expand buffer but raise warning?